driver/pipeline/middleware.h

```diff
@@ -0,0 +1,262 @@
+// Copyright 2024 Synnax Labs, Inc.
+//
+// Use of this software is governed by the Business Source License included in the file
+// licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with the Business Source
+// License, use of this software will be governed by the Apache License, Version 2.0,
+// included in the file licenses/APL.txt.
+
+#pragma once
+
+#include <memory>
+#include <functional>
+#include <vector>
+#include <map>
+#include <string>
+#include <variant>
+#include <thread>
+
+#include "client/cpp/synnax.h"
+#include "driver/config/config.h"
+
+namespace pipeline {
+///////////////////////////////////////////////////////////////////////////////////
+//                                    Middleware                                 //
+///////////////////////////////////////////////////////////////////////////////////
+class Middleware {
+public:
+    virtual ~Middleware() = default;
+
+    virtual bool handle(Frame &frame) = 0;
+}; // class Middleware
+
+///////////////////////////////////////////////////////////////////////////////////
+//                                  MiddlewareChain                              //
+///////////////////////////////////////////////////////////////////////////////////
+class MiddlewareChain {
+public:
+    void add(std::shared_ptr<Middleware> middleware) {
+        middlewares.push_back(middleware);
+    }
+
+    bool empty() {
+        return middlewares.empty();
+    }
+
+    freighter::Error exec(Frame &frame) {
+        for (auto &middleware: middlewares) {
+            if (!middleware->handle(frame)) {
+                return freighter::Error("Middleware failed");
+            }
+        }
+        return freighter::NIL;
+    }
+
+private:
+    std::vector<std::shared_ptr<Middleware> > middlewares;
+}; // class MiddlewareChain
+
+///////////////////////////////////////////////////////////////////////////////////
+//                                  TareMiddleware                               //
+///////////////////////////////////////////////////////////////////////////////////
+///@brief middleware to tare data written to channels based on the last frame processed at the time of taring
+/// This middleware should added to the pipeline middleware chain first so that it can tare the data before any other middleware
+/// can process it.
+class TareMiddleware : public Middleware {
+public:
+    explicit TareMiddleware(std::vector<synnax::ChannelKey> keys) {
+        for (auto &key: keys) {
+            tare_values[key] = 0.0;
+        }
+    }
+
+    void tare(json &arg) {
+        // if json contains no keys, tare everything
```

Comment:
My comment here was resolved and everything looks the same. Why aren't we using a parser here? There's no guarantee that the json you receive will contain keys, and this will throw an exception if there are no keys, which could crash the driver